📖 Document multi-tenancy contract (#4074)
kubernetes-sigs/cluster-api: docs/book/src/developer/architecture/controllers/support-multiple-instances.md

# Support running multiple instances of the same provider

> In releases <= v1alpha3 the need of supporting multiple credentials was addressed by running multiple
> instances of the same provider, each one with its own credentials.

Reviewer: ```suggestion
Up until v1alpha3, the need of supporting multiple credentials was addressed by running multiple
instances of the same provider, each one with its own set of credentials while watching different namespaces.
```

> However, running multiple instances of the same provider proved to be complicated for several reason like e.g.

Reviewer: ```suggestion
However, running multiple instances of the same provider proved to be complicated for several reason:
```

> - Complexity in packaging providers, due to the requirement to split CRDs and the web-hooks from the controller given that
>   only the controller can actually be installed many times, while CRDs and web-hooks can exist only in one instance.

Reviewer: ```suggestion
- Complexity in packaging providers: CustomResourceDefinitions (CRD) are global resources, these may have a reference to a service that can be used to covert between CRD versions (conversion webhooks). Only one of these services should be running at any given time, this requirement led us to previously split the webhooks code to a different deployment and namespace.
```

> - Complexity in deploying providers, due to the requirement to ensure consistency of the management cluster e.g. avoid
>   controllers watching for the same namespaces, etc. 

Reviewer: ```suggestion
- Complexity in deploying providers, due to the requirement to ensure consistency of the management cluster, e.g. controllers watching the same namespaces.
```

- The introduction of the concept of management groups in clusterctl, with impacts on the user experience/documentation.

> - Complexity in managing co-existence of different versions of the same provider while there could be only
>   one version of CRDs and web-hooks. 

Reviewer: ```suggestion
- Complexity in managing co-existence of different versions of the same provider while there could be only
  one version of CRDs and webhooks. 
```

> - A residual risk while running different version of the same provider, because some version of the provider
>   de-facto were forced to run with CRDs and web-hooks deployed from a different version.

Reviewer: ```suggestion
```

Either remove or include in the first point, seems like a duplicate?

Author: re- arranged in one sentence

Nevertheless, we want to make it possible for users to choose to deploy multiple instances of the same providers, 
in case the above limitations/extra complexity are acceptable for them.

## Contract

In order to make it possible for users to deploy multiple instances of the same provider:

> - All the provider will be required to support the `--namespace` flag in their controllers.

Reviewer: ```suggestion
- All the provider MUST support the `--namespace` flag in their controllers.
```

> Users selecting this deployment model should be aware that:

Reviewer: ```suggestion
⚠️  Users selecting this deployment model, please be aware:
```

> - In the Cluster API project, we are not going to include release a `component.yaml` file supporting this scenario; 
>   however it will be possible to create such file starting from the content of the `/config` folder.

Reviewer: ```suggestion
- Support should be considered best-effort.
- Cluster API (incl. every provider managed under `kubernetes-sigs`, won't release a specialized components file supporting the scenario described above; however, users should be able to create such deployment model from the `/config` folder.
```

> - clusterctl and the CAPI provider operator are not going to support deploying/managing multiple instances of the same provider.

Reviewer: ```suggestion
```

Author: I think we should state this in order to set clear expectations, unless you think this is clear from the previous statement

> - The Cluster API testing infrastructure is not going to run test cases with multiple instances of the same provider.

Reviewer: ```suggestion
- Cluster API (incl. every provider managed under `kubernetes-sigs`) testing infrastructure won't run test cases with multiple instances of the same provider.
```

>   

Reviewer: ```suggestion
  
In conclusion, giving the increasingly complex task that is to manage multiple instances of the same controllers, the Cluster API community may only provide best effort support for users that choose this model.

As always, if some members of the community would like to take on the responsibility of managing this model, please reach out through the usual communication channels, we'll make sure to guide you in the right path.
```

# Support running multiple instances of the same provider

> Up until v1alpha3, the need of supporting multiple credentials was addressed by running multiple

Reviewer 2: Might make sense to also add a note to https://github.com/kubernetes-sigs/cluster-api/blob/master/docs/book/src/developer/providers/v1alpha3-to-v1alpha4.md

Author: Done, PTAL

> Up until v1alpha3, the need of supporting multiple credentials was addressed by running multiple

Reviewer 3: I feel like this sentence may be missing something when read by someone new to CAPI. 
That is, the reader may ask _What is the need for supporting multiple credentials?_ or _What does multiple credentials have to do with multi-tenancy of providers?_

But I'm unsure how to word this myself. My thoughts lead me to - multiple credentials required to define workload clusters within separate IaaS accounts.

Reviewer 3: I noticed that in the Glossary section we have 

> Multi tenancy in Cluster API defines the capability of an infrastructure provider to manage different credentials, each one of them corresponding to an infrastructure tenant.

Maybe we just link to the Multi tenancy glossary entry here.

instances of the same provider, each one with its own set of credentials while watching different namespaces.

However, running multiple instances of the same provider proved to be complicated for several reasons:

- Complexity in packaging providers: CustomResourceDefinitions (CRD) are global resources, these may have a reference

>   to a service that can be used to covert between CRD versions (conversion webhooks). Only one of these services should

Reviewer 3: nit:
```suggestion
  to a service that can be used to convert between CRD versions (conversion webhooks). Only one of these services should
```

  be running at any given time, this requirement led us to previously split the webhooks code to a different deployment
  and namespace.
- Complexity in deploying providers, due to the requirement to ensure consistency of the management cluster, e.g. 
  controllers watching the same namespaces. 
- The introduction of the concept of management groups in clusterctl, with impacts on the user experience/documentation.
- Complexity in managing co-existence of different versions of the same provider while there could be only
  one version of CRDs and webhooks. Please note that this constraint generates a risk, because some version of the provider 
  de-facto were forced to run with CRDs and webhooks deployed from a different version.

Nevertheless, we want to make it possible for users to choose to deploy multiple instances of the same providers, 
in case the above limitations/extra complexity are acceptable for them.

## Contract

In order to make it possible for users to deploy multiple instances of the same provider:

> - All the provider MUST support the `--namespace` flag in their controllers.

Reviewer 2: "All the provider" missing word or "s"?

Reviewer 2: ```suggestion
- Providers MUST support the `--namespace` flag in their controllers.
```

maybe
